File: Ejercicio 17/Ejercicio 17.js
// Clase del nodo
class Nodo {
    constructor(v) {
        this.valor = v;
        this.siguiente = null;
        this.anterior = null;
    }
}

// Clase de Lista Circular Doble
class ListaCircularDoble {
    constructor() {
        this.cabeza = null;
        this.tamaño = 0;
    }
    
    // Insertar al final
    insertarFinal(valor) {
        const nuevoNodo = new Nodo(valor);
        
        if (this.cabeza === null) {
            this.cabeza = nuevoNodo;
            this.cabeza.siguiente = this.cabeza;
            this.cabeza.anterior = this.cabeza;
        } else {
            const ultimo = this.cabeza.anterior;
            
            ultimo.siguiente = nuevoNodo;
            nuevoNodo.anterior = ultimo;
            nuevoNodo.siguiente = this.cabeza;
            this.cabeza.anterior = nuevoNodo;
        }
        
        this.tamaño++;
    }
    
    // Mostrar lista en sentido horario
    mostrarHorario() {
        if (this.cabeza === null) {
            console.log("Lista vacía");
            return;
        }
        
        process.stdout.write("Lista en sentido horario: ");
        let actual = this.cabeza;
        
        do {
            process.stdout.write(actual.valor + " ");
            actual = actual.siguiente;
        } while (actual !== this.cabeza);
        
        console.log();
    }
    
    // Mostrar lista en sentido antihorario
    mostrarAntihorario() {
        if (this.cabeza === null) {
            console.log("Lista vacía");
            return;
        }
        
        process.stdout.write("Lista en sentido antihorario: ");
        let actual = this.cabeza.anterior;
        
        do {
            process.stdout.write(actual.valor + " ");
            actual = actual.anterior;
        } while (actual !== this.cabeza.anterior);
        
        console.log();
    }
    
    // Obtener tamaño
    getTamaño() {
        return this.tamaño;
    }
    
    // Generar lista con valores aleatorios
    generarListaAleatoria(cantidad) {
        for (let i = 0; i < cantidad; i++) {
            const valor = Math.floor(Math.random() * 100) + 1; // Valores entre 1 y 100
            this.insertarFinal(valor);
        }
    }
}

// Función principal
function main() {
    const lista = new ListaCircularDoble();
    
    console.log("=== LISTA ENLAZADA CIRCULAR DOBLE ===");
    console.log("Generando 15 valores aleatorios");
    console.log();
    
    // Generar lista con 15 valores aleatorios
    lista.generarListaAleatoria(15);
    
    // Mostrar resultados
    console.log("Tamaño de la lista: " + lista.getTamaño());
    
    lista.mostrarHorario();
    lista.mostrarAntihorario();
}

// Versión alternativa con console.log para navegadores
class ListaCircularDobleBrowser {
    constructor() {
        this.cabeza = null;
        this.tamaño = 0;
    }
    
    // Insertar al final (igual que antes)
    insertarFinal(valor) {
        const nuevoNodo = new Nodo(valor);
        
        if (this.cabeza === null) {
            this.cabeza = nuevoNodo;
            this.cabeza.siguiente = this.cabeza;
            this.cabeza.anterior = this.cabeza;
        } else {
            const ultimo = this.cabeza.anterior;
            
            ultimo.siguiente = nuevoNodo;
            nuevoNodo.anterior = ultimo;
            nuevoNodo.siguiente = this.cabeza;
            this.cabeza.anterior = nuevoNodo;
        }
        
        this.tamaño++;
    }
    
    // Mostrar lista en sentido horario (versión para navegador)
    mostrarHorario() {
        if (this.cabeza === null) {
            console.log("Lista vacía");
            return;
        }
        
        let resultado = "Lista en sentido horario: ";
        let actual = this.cabeza;
        
        do {
            resultado += actual.valor + " ";
            actual = actual.siguiente;
        } while (actual !== this.cabeza);
        
        console.log(resultado);
    }
    
    // Mostrar lista en sentido antihorario (versión para navegador)
    mostrarAntihorario() {
        if (this.cabeza === null) {
            console.log("Lista vacía");
            return;
        }
        
        let resultado = "Lista en sentido antihorario: ";
        let actual = this.cabeza.anterior;
        
        do {
            resultado += actual.valor + " ";
            actual = actual.anterior;
        } while (actual !== this.cabeza.anterior);
        
        console.log(resultado);
    }
    
    // Obtener tamaño
    getTamaño() {
        return this.tamaño;
    }
    
    // Generar lista con valores aleatorios
    generarListaAleatoria(cantidad) {
        for (let i = 0; i < cantidad; i++) {
            const valor = Math.floor(Math.random() * 100) + 1;
            this.insertarFinal(valor);
        }
    }
}
main();

File: Ejercicio 17/Ejercicio 17.cs
using System;

// Clase del nodo
public class Nodo 
{
    public int valor;
    public Nodo siguiente;
    public Nodo anterior;
    
    public Nodo(int v) 
    {
        valor = v;
        siguiente = null;
        anterior = null;
    }
}

// Clase de Lista Circular Doble
public class ListaCircularDoble 
{
    private Nodo cabeza;
    private int tamaño;
    
    // Constructor
    public ListaCircularDoble() 
    {
        cabeza = null;
        tamaño = 0;
    }
    
    // Insertar al final
    public void InsertarFinal(int valor) 
    {
        Nodo nuevoNodo = new Nodo(valor);
        
        if (cabeza == null) 
        {
            cabeza = nuevoNodo;
            cabeza.siguiente = cabeza;
            cabeza.anterior = cabeza;
        } 
        else 
        {
            Nodo ultimo = cabeza.anterior;
            
            ultimo.siguiente = nuevoNodo;
            nuevoNodo.anterior = ultimo;
            nuevoNodo.siguiente = cabeza;
            cabeza.anterior = nuevoNodo;
        }
        
        tamaño++;
    }
    
    // Mostrar lista en sentido horario
    public void MostrarHorario() 
    {
        if (cabeza == null) 
        {
            Console.WriteLine("Lista vacía");
            return;
        }
        
        Console.Write("Lista en sentido horario: ");
        Nodo actual = cabeza;
        
        do 
        {
            Console.Write(actual.valor + " ");
            actual = actual.siguiente;
        } while (actual != cabeza);
        
        Console.WriteLine();
    }
    
    // Mostrar lista en sentido antihorario
    public void MostrarAntihorario() 
    {
        if (cabeza == null) 
        {
            Console.WriteLine("Lista vacía");
            return;
        }
        
        Console.Write("Lista en sentido antihorario: ");
        Nodo actual = cabeza.anterior;
        
        do 
        {
            Console.Write(actual.valor + " ");
            actual = actual.anterior;
        } while (actual != cabeza.anterior);
        
        Console.WriteLine();
    }
    
    // Obtener tamaño
    public int GetTamaño() 
    {
        return tamaño;
    }
    
    // Generar lista con valores aleatorios
    public void GenerarListaAleatoria(int cantidad) 
    {
        Random rand = new Random();
        
        for (int i = 0; i < cantidad; i++) 
        {
            int valor = rand.Next(1, 101); // Valores entre 1 y 100
            InsertarFinal(valor);
        }
    }
}

// Clase principal
public class Program 
{
    public static void Main(string[] args) 
    {
        ListaCircularDoble lista = new ListaCircularDoble();
        
        Console.WriteLine("=== LISTA ENLAZADA CIRCULAR DOBLE ===");
        Console.WriteLine("Generando 15 valores aleatorios");
        Console.WriteLine();
        
        // Generar lista con 15 valores aleatorios
        lista.GenerarListaAleatoria(15);
        
        // Mostrar resultados
        Console.WriteLine("Tamaño de la lista: " + lista.GetTamaño());
        
        lista.MostrarHorario();
        lista.MostrarAntihorario();
    }
}